cpp/src/strings/filling/fill.cu

```diff
@@ -27,35 +25,34 @@
 #include <rmm/resource_ref.hpp>
 
 #include <thrust/iterator/counting_iterator.h>
+#include <thrust/transform.h>
 
 namespace cudf {
 namespace strings {
 namespace detail {
 namespace {
+
 struct fill_fn {
   column_device_view const d_strings;
   size_type const begin;
   size_type const end;
-  string_view const d_value;
-  size_type* d_offsets{};
-  char* d_chars{};
-
-  __device__ string_view resolve_string_at(size_type idx) const
-  {
-    if ((begin <= idx) && (idx < end)) { return d_value; }
-    return d_strings.is_valid(idx) ? d_strings.element<string_view>(idx) : string_view{};
-  }
+  string_scalar_device_view const d_value;
 
-  __device__ void operator()(size_type idx) const
+  __device__ string_index_pair operator()(size_type idx) const
   {
-    auto const d_str = resolve_string_at(idx);
-    if (!d_chars) {
-      d_offsets[idx] = d_str.size_bytes();
+    auto d_str = string_view();
+    if ((begin <= idx) && (idx < end)) {
+      if (!d_value.is_valid()) { return string_index_pair{nullptr, 0}; }
```

Comment:
Since `d_value` is const, would it be more feasible to check `d_value.is_valid()` just once before calling `transform` instead of here for each `idx`?

Reply:
Calling `is_valid()` on the `scalar_device_view` in host code is actually a device-to-host copy including a stream sync. 
Calling `is_valid()` here is very cheap since it is just returning a boolean and is likely inlined. 
It also should not incur any divergence (over and above the [begin,end) range divergrence) since all threads will have the same value.

Reply:
Sounds good!